### PR TITLE
[DPE-5588] Check against invalid arch

### PR DESCRIPTION
**Pull request by sinclert-canonical**
This PR introduces the usage of the newly created architecture module, similar to the one in the MySQL lib (see [code](https://github.com/canonical/mysql-operator/blob/main/lib/charms/mysql/v0/architecture.py)).

The shape of the integration tests, as well as the helper used to get the built charm in the architecture that mismatches the one of the host itself, have been adapted from the one Lucas created for PostgreSQL-k8s (see [code](https://github.com/canonical/postgresql-k8s-operator/blob/70a96ce33eb0de054b9297e3bd558f6ece54c0b0/tests/integration/test_wrong_arch.py)).

### Additional info

- The new logic has been added as a module, and not as a charm lib, due to @carlcsaposs-canonical preference.
- The integration test fixtures have been defined in `conftest.py` and not in `markers.py` to respect the existing setup.
- The integration test _base_ indexes have been tweaked, to account for [this legacy base](https://github.com/canonical/mysql-router-operator/blob/d0af04794698df4ae2a5166f30607ebb1d7cd975/charmcraft.yaml#L9-L11). Should we remove it?

### CODE REVIEW

**Comment by sinclert-canonical** [comment]
The proposed checks fail because of the number of units on a _subordinate charm_ is `0`, but I originally set it to `1` and failed nonetheless. Should I deploy MySQL beforehand for the router to fully deploy?

File: tests/conftest.py [hunk 1]
```diff
@@ -32,6 +33,30 @@ def pytest_configure(config):
...
+        check=True,
+        encoding="utf-8",
+    ).stdout.strip()
+
+
+@pytest.fixture
+def only_amd64(architecture):
+    """Pretty way to skip ARM tests."""
+    if architecture != "amd64":
+        pytest.skip("Requires amd64 architecture")
```
[the remark this answers is not in the document]
Comment (sinclert-canonical) [reply tests/conftest.py:50]: This repo already has _markers_ such as `only_with_juju_secrets` and `only_without_juju_secrets`, which you could find in the `markers.py` module of sibling repositories (see [mysql-operator](https://github.com/canonical/mysql-operator/blob/main/tests/integration/markers.py) and [mysql-k8s-operator](https://github.com/canonical/mysql-k8s-operator/blob/main/tests/integration/markers.py)), defined in this file.

I would say if you want them to be moved to a `markers.py` module to be more consistent with its K8s counterpart, we should also move the existing markers, and not only the ones I am proposing. At that point, the effort lays more into the _refactor_ category and IMHO deserves its own separate PR.

File: tests/unit/test_architecture.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from unittest.mock import patch
```
[the remark this answers is not in the document]
Comment (sinclert-canonical) [reply tests/unit/test_architecture.py]: Done in commit https://github.com/canonical/mysql-router-operator/pull/194/commits/9b64f258baa9b08ed01ee43799f5fca2ae77d4c2.

**Review by taurus-forever — APPROVED** [review]
LGTM from my side, please sync with Carl directly. Tnx!

**Comment by sinclert-canonical** [comment]
📢 We need to limit the execution of the integration tests to Ubuntu 22.04 (Jammy), as there is no ARM64 version of Ubuntu 20.04 (Focal), currently supported by this charm (see [charmcraft.yaml](https://github.com/canonical/mysql-router-operator/blob/d0af04794698df4ae2a5166f30607ebb1d7cd975/charmcraft.yaml#L5-L17)).

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**